### PR TITLE
juliagong/add_lp_error_evaluator

### PR DESCRIPTION
**Pull request by juliagongms**

### CODE REVIEW

**Comment by msxingpeng** [comment]
We usually submit PR to this repo from our own folk. See an example here: https://github.com/microsoft/vision-evaluation/pull/32

File: vision_evaluation/evaluators.py [hunk 1]
```diff
@@ -1100,3 +1100,46 @@ def add_predictions(self, predictions, ground_truths):
 
     def get_report(self, **kwargs):
         return {self._get_id(): {group: eval.get_report(**kwargs) for group, eval in self._evaluators.items()}}
+
+
+class MeanLpErrorEvaluator(Evaluator):
+    """
+    Error evaluator for regression targets
+    Parameters:
```
Comment (msxingpeng) [inline vision_evaluation/evaluators.py:1108]: The parameters doc should be in the __init__ function

File: vision_evaluation/evaluators.py [hunk 2]
```diff
@@ -1100,3 +1100,46 @@ def add_predictions(self, predictions, ground_truths):
...
+
+        super(MeanLpErrorEvaluator, self).__init__()
+        
+    def reset(self):
+        super(MeanLpErrorEvaluator, self).reset()
+        self.total_num = 0
+        self.total_error = 0
+
+    def _get_id(self):
+        return f'l{self.p}'
```
Comment (msxingpeng) [inline vision_evaluation/evaluators.py:1127]: return f'MeanL{self.p}Error'

File: vision_evaluation/evaluators.py [hunk 3]
```diff
@@ -1100,3 +1100,46 @@ def add_predictions(self, predictions, ground_truths):
...
+
+    def add_predictions(self, predictions, targets):
+        """ Evaluate a batch of predictions.
+        Args:
+            predictions: the model output numpy array. Shape (N,)
+            targets: the golden truths. Shape (N,)
```
Comment (msxingpeng) [inline vision_evaluation/evaluators.py:1133]: Curious what is the "golden truths" referring to? Ground truth?

File: vision_evaluation/evaluators.py [hunk 4]
```diff
@@ -1100,3 +1100,46 @@ def add_predictions(self, predictions, ground_truths):
...
+        assert len(targets.shape) == 1
+
+        n_sample = len(predictions)
+
+        self.total_error += np.linalg.norm(predictions - targets, ord=self.p)**self.p
+        self.total_num += n_sample
+    
+    def get_report(self, **kwargs):
+        return {f'{self._get_id()}': (float(self.total_error)**(1/self.p) / self.total_num) if self.total_num else 0.0}
+
```
Comment (msxingpeng) [inline vision_evaluation/evaluators.py:1145]: check flake8 here

File: vision_evaluation/evaluators.py [hunk 5]
```diff
@@ -1100,3 +1100,46 @@ def add_predictions(self, predictions, ground_truths):
...
+class MeanLpErrorEvaluator(Evaluator):
+    """
+    Error evaluator for regression targets
+    Parameters:
+        p: Lp order of the norm used
+    """
+
+    def __init__(self, p: int):
+        assert p != 0
```
Comment (pj-ms) [inline vision_evaluation/evaluators.py:1113]: P > 0?

File: vision_evaluation/evaluators.py (hunk 2 above)
Comment (pj-ms) [inline vision_evaluation/evaluators.py:1127]: f'l{self.p}_error'

File: vision_evaluation/evaluators.py [hunk 6]
```diff
@@ -1100,3 +1100,46 @@ def add_predictions(self, predictions, ground_truths):
...
+        Args:
+            predictions: the model output numpy array. Shape (N,)
+            targets: the golden truths. Shape (N,)
+        """
+        assert len(predictions) == len(targets)
+        assert len(targets.shape) == 1
+
+        n_sample = len(predictions)
+
+        self.total_error += np.linalg.norm(predictions - targets, ord=self.p)**self.p
```
Comment (pj-ms) [inline vision_evaluation/evaluators.py:1140]: why not

sum(np.power(predictions - targets), self.p))

File: vision_evaluation/evaluators.py [hunk 7]
```diff
@@ -1100,3 +1100,46 @@ def add_predictions(self, predictions, ground_truths):
...
+
+
+class MeanLpErrorEvaluator(Evaluator):
+    """
+    Error evaluator for regression targets
+    Parameters:
+        p: Lp order of the norm used
+    """
+
+    def __init__(self, p: int):
```
Comment (pj-ms) [inline vision_evaluation/evaluators.py:1112]: you have a test of p = 0.5

**Review by pj-ms — CHANGES_REQUESTED** [review]

File: vision_evaluation/evaluators.py (hunk 5 above)
Comment (juliagongms) [reply vision_evaluation/evaluators.py:1113]: Good point - I was originally thinking p < 0 is ok, but I don't see where we'd use that. Changed.

File: vision_evaluation/evaluators.py (hunk 1 above)
Comment (juliagongms) [reply vision_evaluation/evaluators.py:1108]: Good catch, thanks!

File: vision_evaluation/evaluators.py (hunk 7 above)
Comment (juliagongms) [reply vision_evaluation/evaluators.py:1112]: Yes, theoretically that could be an Lp norm that someone may want to calculate, so I was thinking to leave that option open? Your thoughts?

File: vision_evaluation/evaluators.py (hunk 2 above)
Comment (juliagongms) [reply vision_evaluation/evaluators.py:1127]: Thanks! Modified slightly to follow the format of the other metrics; I used lowercase + underscores.

File: vision_evaluation/evaluators.py (hunk 3 above)
Comment (juliagongms) [reply vision_evaluation/evaluators.py:1133]: Thanks for catching this - I copy-pasted this section from some other doc comments that someone else wrote. Changed for clarity.

File: vision_evaluation/evaluators.py (hunk 6 above)
Comment (juliagongms) [reply vision_evaluation/evaluators.py:1140]: Good point, added np.abs also. Thanks!

File: vision_evaluation/evaluators.py (hunk 7 above)
Comment (juliagongms) [reply vision_evaluation/evaluators.py:1110]: Test case removed since it's not commonly used.

File: vision_evaluation/evaluators.py [hunk 8]
```diff
@@ -1100,3 +1100,47 @@ def add_predictions(self, predictions, ground_truths):
 
     def get_report(self, **kwargs):
         return {self._get_id(): {group: eval.get_report(**kwargs) for group, eval in self._evaluators.items()}}
+
+
+class MeanLpErrorEvaluator(Evaluator):
+    """
+    Error evaluator for regression targets
```
Comment (pj-ms) [inline vision_evaluation/evaluators.py:1107]: Mean LP error

**Review by pj-ms — APPROVED** [review]

**Pull request merged**